File: ProjectEuler/p9.cc
//author agis daniels
//goal is What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
#include <iostream> // cin cout 
#include <math.h> // sqrt 

using namespace std;

long gcd(long a, long b) {
    return b == 0 ? a : gcd(b, a % b);
}

int main(){
    long s = 1000;
    long s2 = s/2;
    long mlim = ceil(sqrt(s2))-1;
    for(long m=2; m<=mlim; ++m){
        if(0==(s2%m)){
            long sm=s2/m;
            for(; 0==(sm%2); sm/=2){}
            long k;
            if(1==(m%2)){
                k=m+2;
            }else{
                k=m+1;
            }
            while((k<m+m)&&(k<=sm)){
                if(0==(sm%k)&&(1==gcd(k, m))){
                    long d = s2/(k*m);
                    long n = k-m;
                    long a = d*(m*m-n*n);
                    long b = d*m*n+d*m*n;
                    long c = d*(m*m+n*n);
                    cout << a*c*b <<endl;
                    cout << (a+c+b==s) <<endl;
                    cout << s <<endl;
                    cout << a <<endl;
                    cout << b <<endl;
                    cout << c <<endl;
                }
                 k+=2;
            }
        }
    }
    return 0;
}

File: ProjectEuler/Python/p9.py
from math import sqrt, ceil

def gcd(a, b):
  return a if b == 0 else gcd(b, a % b)


def sol1(lim):
  for a in range(3, (lim-3)//3):
    for b in range(a+1, (lim-1-a)//2):
      c = lim - a - b
      if c*c == b*b + a*a:
        print('the triplet where a + b + c == 1000 and a^2 + b^2 = c^2 is {}'.format(a*b*c))

def sol2(lim):
  ''' a studied solution '''
  s2 = lim//2
  mlim = int(ceil(sqrt(s2))-1)
  for m in range(2, mlim + 1):
    if 0 == s2%m:
      sm = s2//m
      while 0 == sm%2:
        sm //= 2
      k = 0
      if 1 == m%2:
        k = m + 2
      else:
        k = m + 1
      while k < m+m and k <= sm:
        if 0 == sm%k and 1 == gcd(k, m):
          d = s2 // (k*m)
          n = k-m
          a = d * (m*m - n*n)
          b = d*m*n*2
          c = d * (m*m + n*n)
          print('the triplet where a + b + c == 1000 and a^2 + b^2 = c^2 is {}'.format(a*b*c))
        k += 2

def main():
  sol2(1000)
  sol1(1000)
main()
 
